toolchain/semantics/semantics_node.h

```diff
@@ -0,0 +1,129 @@
+      : identifier(identifier) {}
+  explicit SemanticsNodeArgs(SemanticsIntegerLiteralId integer_literal)
+      : integer_literal(integer_literal) {}
+
+  int no_args[0];
+  SemanticsNodeId one_node;
+  SemanticsTwoNodeIds two_nodes;
+  SemanticsIdentifierId identifier;
+  SemanticsIntegerLiteralId integer_literal;
+};
+static_assert(sizeof(SemanticsNodeArgs) == 8, "Unexpected OneOfArgs size");
+
+// The standard structure for nodes.
+class SemanticsNode {
+ public:
+  // Define factory functions for each node kind. These should improve type
+  // safety by enforcing argument counts.
+#define CARBON_SEMANTICS_MAKE_no_args(Name)                               \
+  static auto Make##Name()->SemanticsNode {                               \
+    return SemanticsNode(SemanticsNodeKind::Name(), SemanticsNodeArgs()); \
+  }
+#define CARBON_SEMANTICS_MAKE_one_node(Name)                        \
+  static auto Make##Name(SemanticsNodeId one_node)->SemanticsNode { \
+    return SemanticsNode(SemanticsNodeKind::Name(),                 \
+                         SemanticsNodeArgs(one_node));              \
+  }
+#define CARBON_SEMANTICS_MAKE_two_nodes(Name)                          \
+  static auto Make##Name(SemanticsNodeId node1, SemanticsNodeId node2) \
+      ->SemanticsNode {                                                \
+    return SemanticsNode(                                              \
+        SemanticsNodeKind::Name(),                                     \
+        SemanticsNodeArgs(SemanticsTwoNodeIds{node1, node2}));         \
+  }
+#define CARBON_SEMANTICS_MAKE_identifier(Name)                              \
+  static auto Make##Name(SemanticsIdentifierId identifier)->SemanticsNode { \
+    return SemanticsNode(SemanticsNodeKind::Name(),                         \
+                         SemanticsNodeArgs(identifier));                    \
+  }
+#define CARBON_SEMANTICS_MAKE_integer_literal(Name)                 \
+  static auto Make##Name(SemanticsIntegerLiteralId integer_literal) \
+      ->SemanticsNode {                                             \
+    return SemanticsNode(SemanticsNodeKind::Name(),                 \
+                         SemanticsNodeArgs(integer_literal));       \
+  }
+
+#define CARBON_SEMANTICS_NODE_KIND(Name, ArgsType) \
+  CARBON_SEMANTICS_MAKE_##ArgsType(Name)
+#include "toolchain/semantics/semantics_node_kind.def"
+
+#undef CARBON_SEMANTICS_MAKE_no_args
+#undef CARBON_SEMANTICS_MAKE_one_node
+#undef CARBON_SEMANTICS_MAKE_two_nodes
+#undef CARBON_SEMANTICS_MAKE_identifier
+#undef CARBON_SEMANTICS_MAKE_integer_literal
+
+  SemanticsNode() : kind_(SemanticsNodeKind::Invalid()) {}
+
+  auto kind() -> SemanticsNodeKind { return kind_; }
+
+  void Print(llvm::raw_ostream& out) const;
+
+ private:
+  SemanticsNode(SemanticsNodeKind kind, SemanticsNodeArgs one_of_args)
+      : kind_(kind), one_of_args_(one_of_args) {}
+
+  SemanticsNodeKind kind_;
+
+  SemanticsNodeArgs one_of_args_;
```

Comment:
(FWIW: I bet there are few enough two-operand nodes that we can steal a few bits from the node IDs and fit this into 8 bytes rather than 12. But I'm happy not doing that in this change.)

Reply:
Added a note.